streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -49,48 +52,233 @@ public StreamTask restoringTaskFor(final TopicPartition partition) {
         final List<StreamTask> tasks = super.allTasks();
         tasks.addAll(restoring.values());
+        tasks.addAll(suspended.values());
         return tasks;
     }
 
     @Override
     Set<TaskId> allAssignedTaskIds() {
         final Set<TaskId> taskIds = super.allAssignedTaskIds();
         taskIds.addAll(restoring.keySet());
+        taskIds.addAll(suspended.keySet());
         return taskIds;
     }
 
     @Override
     boolean allTasksRunning() {
-        return super.allTasksRunning() && restoring.isEmpty();
+        return super.allTasksRunning() && restoring.isEmpty() && suspended.isEmpty();
     }
 
-    RuntimeException closeAllRestoringTasks() {
+    @Override
+    void closeTask(final StreamTask task, final boolean clean) {
+        if (suspended.containsKey(task.id())) {
+            task.closeSuspended(clean, false, null);
+        } else {
+            task.close(clean, false);
+        }
+    }
+    
+    Set<TaskId> suspendedTaskIds() {
+        return suspended.keySet();
+    }
+
+    RuntimeException revokeTasks(final Set<TaskId> revokedTasks, final List<TopicPartition> revokedChangelogs) {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+        final Set<TaskId> revokedRunningTasks = new HashSet<>();
+        final Set<TaskId> revokedNonRunningTasks = new HashSet<>();
+        final Set<TaskId> revokedRestoringTasks = new HashSet<>();
+
+        for (final TaskId task : revokedTasks) {
+            if (running.containsKey(task)) {
+                revokedRunningTasks.add(task);
+            } else if (created.containsKey(task)) {
+                revokedNonRunningTasks.add(task);
+            } else if (restoring.containsKey(task)) {
+                revokedRestoringTasks.add(task);
+            } else {
+                log.error("Task {} was revoked but cannot be found in the assignment", task);
+                firstException.set(new IllegalStateException("Revoked a task that was not in the assignment"));
+            }
+        }
+
+        firstException.compareAndSet(null, suspendRunningTasks(revokedRunningTasks, revokedChangelogs));
+        firstException.compareAndSet(null, closeNonRunningTasks(revokedNonRunningTasks, revokedChangelogs));
+        firstException.compareAndSet(null, closeRestoringTasks(revokedRestoringTasks, revokedChangelogs));
+
+        return firstException.get();
+    }
+
+    private RuntimeException suspendRunningTasks(final Set<TaskId> revokedRunningTasks, final List<TopicPartition> revokedChangelogs) {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+        for (final Iterator<TaskId> it = revokedRunningTasks.iterator(); it.hasNext(); ) {
+            final TaskId id = it.next();
+            final StreamTask task = running.get(id);
+
+            // if the task is already suspended just log a message and move on
+            if (suspended.containsKey(id)) {
```

Comment:
As discussed offline, the suspended and running sets can overlap so we can use `running` to report active tasks consistently for cooperative and eager. But I realize in my second PR this is not needed as in cooperative we should not need to encode active tasks since we have the `ownedPartitions`. 
So, I think I will revert back to using the `prevActiveTasks`